pallets/external-validators-rewards/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+//! This pallet keep tracks of the validators reward points.
+//! Storage will be cleared after a period of time.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+
+use {
+    frame_support::traits::{Defensive, Get, ValidatorSet},
+    polkadot_primitives::ValidatorIndex,
+    runtime_parachains::session_info,
+    sp_staking::SessionIndex,
+    sp_std::collections::btree_set::BTreeSet,
+};
+
+#[frame_support::pallet]
+pub mod pallet {
+    use {
+        frame_support::pallet_prelude::*, sp_std::collections::btree_map::BTreeMap,
+        tp_traits::EraIndexProvider,
+    };
+
+    /// The current storage version.
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+    pub type RewardPoints = u32;
+    pub type EraIndex = u32;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// How to fetch the current era info.
+        type EraIndexProvider: EraIndexProvider;
+
+        /// For how many eras points are kept in storage.
+        #[pallet::constant]
+        type HistoryDepth: Get<EraIndex>;
+
+        /// The amount of era points given by backing a candidate that is included.
+        #[pallet::constant]
+        type BackingPoints: Get<u32>;
+
+        /// The amount of era points given by dispute voting on a candidate.
+        #[pallet::constant]
+        type DisputeStatementPoints: Get<u32>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    /// Keep tracks of distributed points per validator and total.
+    #[derive(RuntimeDebug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+    pub struct EraRewardPoints<AccountId> {
+        pub total: RewardPoints,
+        pub individual: BTreeMap<AccountId, RewardPoints>,
+    }
+
+    impl<AccountId> Default for EraRewardPoints<AccountId> {
+        fn default() -> Self {
+            EraRewardPoints {
+                total: Default::default(),
+                individual: BTreeMap::new(),
+            }
+        }
+    }
+
+    /// Store reward points per era.
+    /// Note: EraRewardPoints is actually bounded by the amount of validators.
+    #[pallet::storage]
+    #[pallet::unbounded]
+    pub type RewardPointsForEra<T: Config> =
+        StorageMap<_, Twox64Concat, EraIndex, EraRewardPoints<T::AccountId>, ValueQuery>;
+
+    impl<T: Config> Pallet<T> {
+        pub fn reward_by_ids(points: impl IntoIterator<Item = (T::AccountId, RewardPoints)>) {
+            let active_era = T::EraIndexProvider::active_era();
+
+            RewardPointsForEra::<T>::mutate(active_era.index, |era_rewards| {
+                for (validator, points) in points.into_iter() {
+                    *era_rewards.individual.entry(validator).or_default() += points;
+                    era_rewards.total += points;
```

Comment:
How often is this function executed? Once per block or once per era? Or something in between?